Fixed upload of large packages

tobias-kuendig opened the pull request:
Since every zip file starts with `PK` we can use this to match the regex against.

ikkentim commented:
This might be a difference between different NuGet clients, I'll test if my NuGet client also works properly with this implementation.

Thanks  :)

tobias-kuendig commented:
You're right. I might revert the changes to prevent breaking changes. I'm currently trying to fix a bug with larger files. It seems like the regex does not match the whole string which leads to truncated zip files. 

A problem that exists definitely is that sometimes the `Content-Type: application/octet-stream` is prepended to the zip archive which makes it invalid and in the second regex the "name" includes the quotes which breaks the name check a few lines later. 

I'll update this PR as soon as I found a solution to all these problems.

ikkentim commented:
Great, thanks :) I've never had to deal with big files. I guess that is at the point the file starts to be uploaded as a multi-part stream

tobias-kuendig committed:
Fixed upload of large packages

In case a large file is uploaded the pcre.backtrack_limit might
not be high enough to get all the file contents via preg_match.
To circumvent this problem we match against a substring of the
request and capture the offsets (PREG_OFFSET_CAPTURE) so we can
later use the captured offset and substr to get the file's contents.

tobias-kuendig committed:
Fixed regex to exclude quotes around the package name

tobias-kuendig commented:
The PR has been updated. 

When uploading large files the `preg_match` call won't work if the `pcre.backtrack_limit` setting is less than the binary string's length. The default setting is `10000000`, so around 10 MB.

To circumvent this problem we can match against a substring of the request and capture the offsets (PREG_OFFSET_CAPTURE) so we can later use the captured offset and substr to get the file's contents.

Also a small bug has been fixed where the package's name was machted including the `"quotes"`.

ikkentim commented:
Thanks! I'll give it a quick test and then merge it

ikkentim commented:
I know it's been ages, but I've tested this just now, and `getFileBlock` seems to returns `null` when I try to push a test package with the latest NuGet client.
